lib/src/logging.rs

```diff
@@ -0,0 +1,37 @@
+use std::collections::HashMap;
+use std::sync::atomic::{AtomicBool, Ordering};
+
+/// Set to true if we failed to write to the journal once
+static EMITTED_JOURNAL_ERROR: AtomicBool = AtomicBool::new(false);
+
+/// Wrapper for systemd structured logging which only emits a message
+/// if we're targeting the system repository, and it's booted.
+pub(crate) fn system_repo_journal_send<K, V>(
+    repo: &ostree::Repo,
+    priority: libsystemd::logging::Priority,
+    msg: &str,
+    vars: impl Iterator<Item = (K, V)>,
+) where
+    K: AsRef<str>,
+    V: AsRef<str>,
+{
+    if !repo.is_system() {
+        return;
+    }
+    if let Err(e) = libsystemd::logging::journal_send(priority, msg, vars) {
```

Comment:
I think you may be interested in https://docs.rs/libsystemd/0.5.0/libsystemd/logging/fn.connected_to_journal.html here. Although I'm not sure if RHEL 8 supports that.

Reply:
In this case I think it's actually a desired feature that we log to the journal even if directly executed from a root shell (i.e. not in a systemd unit).